ext/random/random.stub.php

```diff
@@ -0,0 +1,137 @@
+
+    /** @alias mt_getrandmax */
+    function getrandmax(): int {}
+
+    /** @refcount 1 */
+    function random_bytes(int $length): string {}
+
+    function random_int(int $min, int $max): int {}
+}
+
+namespace Random\Engine
+{
+    /**
+     * @strict-properties
+     */
+    final class Mt19937 implements \Random\Engine
+    {
+        public function __construct(int|null $seed = null, int $mode = MT_RAND_MT19937) {}
+
+        public function generate(): string {}
+
+        public function __serialize(): array {}
+
+        public function __unserialize(array $data): void {}
+
+        public function __debugInfo(): array {}
+    }
+
+    /**
+     * @strict-properties
+     */
+    final class PcgOneseq128XslRr64 implements \Random\Engine
+    {
+        public function __construct(string|int|null $seed = null, string|int $seequence = 0) {}
+
+        /** @implementation-alias Random\Engine\Mt19937::generate */
+        public function generate(): string {}
+
+        public function jump(int $advance): void {}
+        
+        /** @implementation-alias Random\Engine\Mt19937::__serialize */
+        public function __serialize(): array {}
+
+        /** @implementation-alias Random\Engine\Mt19937::__unserialize */
+        public function __unserialize(array $data): void {}
+
+        /** @implementation-alias Random\Engine\Mt19937::__debugInfo */
+        public function __debugInfo(): array {}
+    }
+
+    /**
+     * @strict-properties
+     */
+    final class Xoshiro256StarStar implements \Random\Engine
+    {
+        public function __construct(string|int|null $seed = null) {}
+
+        /** @implementation-alias Random\Engine\Mt19937::generate */
+        public function generate(): string {}
+        
+        public function jump(): void {}
+        
+        public function jumpLong(): void {}
+        
+        /** @implementation-alias Random\Engine\Mt19937::__serialize */
+        public function __serialize(): array {}
+        
+        /** @implementation-alias Random\Engine\Mt19937::__unserialize */
+        public function __unserialize(array $data): void {}
+        
+        /** @implementation-alias Random\Engine\Mt19937::__debugInfo */
+        public function __debugInfo(): array {}
+    }
+
+    /**
+     * @strict-properties
+     * @not-serializable
+     */
+    final class Secure implements \Random\CryptoSafeEngine
+    {
+        /** @implementation-alias Random\Engine\Mt19937::generate */
+        public function generate(): string {}
+    }
+}
+
+namespace Random
+{
+    interface Engine
+    {
+        public function generate(): string;
+    }
+
+    interface CryptoSafeEngine extends Engine
+    {
+    }
+
+    /**
+     * @strict-properties
+     */
+    final class Randomizer
+    {
+        public readonly Engine $engine;
+
+        public function __construct(?Engine $engine = null) {}
+
+        public function getInt(int $min = UNKNOWN, int $max = UNKNOWN): int {}
```

Comment:
Proper default values should be used instead of `UNKNOWN`. E.g.:

```suggestion
        public function getInt(?int $min = null, ?int $max = null): int {}
```

Reply:
This change is not compatible. `getInt(int $min = UNKNOWN, int $max = UNKNOWN): int` accepts "0 or 2" arguments.
The new signature would be contrary to the RFC description and would be difficult to change now...

Reply:
You can validate the number of arguments later. For example `touch()`, we have an error message like this:
>Argument #2 ($mtime) cannot be null when argument #3 ($atime) is an integer

Reply:
OK, I will do that.

Reply:
@kocsismate: Sorry for chiming in but this method is supposed to be a "replacement" for https://github.com/php/php-src/blob/56804e32216574c66cf71359a8a8830e7badc757/ext/standard/basic_functions.stub.php#L1677 so what's wrong with that signature? 😕

Reply:
I would prefer to keep a fully compatible form for once on this. A change to a more appropriate implementation would be good for the next improvement (PHP 8.3?).

However, if `mt_rand()` is modified in PHP 8.2, it will follow.

Reply:
Having `UNKNOWN` parameters has multiple issues:
- the default value is not available for reflection
- the function is not fully compatible with named parameters
- we cannot autogenerate the method signature for the documentation

However, I didn't realize that this method is a replacement of `mt_rand()`. IIRC we didn't fix the signature of `mt_rand()` because its signature is intrinsically overloaded (because it either calls `php_mt_rand_range()` (which takes 2 params) or `php_mt_rand()` (which takes 0), and unfortunately it's the same in case of the `getInt()` method. Then validating the number of arguments like `touch()` may not be the best idea as well. @cmb69 What do you think?

Reply:
Yes, I would like to

1. implement in PHP 8.2 in its current state (`mt_rand()` fully compatible)
2. split into `getInt()` and `nextInt()` (example) in PHP-next (8.3?)

I think it would be appropriate.

It would be a good idea to make the change, but publishing a new RFC now will not be ready in time for the 8.2 Feature Freeze.

However, it may be possible to add new questions to the improved RFC currently being voted on. (I am not sure if that will be allowed or not)

https://wiki.php.net/rfc/random_extension_improvement

Reply:
Overloaded functions are a mess, and we should get rid of them in the long run. And we shouldn't introduce new overloaded functions. In this case this is a bit tricky; amending the RFC in voting is probably not a good idea, and there is no time for a new RFC. However, changing the function signature in 8.3 is also a bad idea wrt. BC. Maybe there is no need for an RFC in this case, and we can just change it prior to PHP 8.2.0, what might best be done via a separate PR.

Reply:
@zeriyoshi I love your idea of having 2 separate methods, makes perfect sense to me! And I also agree with Christoph, none of the options are sensible, so we should change this right now (@ramsey @saundefined @adoy ). I cannot imagine that anyone would resist against having an actual improvement to your RFC which doesn't change things fundamentally.

P.S: congrats for the RFC, it seems like a very nice improvement, and the API is in line what I'd like to see more often in PHP :)

Reply:
I think it can't be just done as a quick follow up PR. In that sense this still follows the RFC - we can merge it immediately and the follow up PR doesn't require RFC or RM exception if there are no objections ofc.

Reply:
For the reference: The follow-up is at #9052.